core/capabilities/ccip/common/extradatacodecregistry.go

```diff
@@ -0,0 +1,93 @@
+package common
+
+import (
+	"sync"
+
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+)
+
+// Ensure ExtraDataCodecRegistry implements the ExtraDataCodecBundle interface from chainlink-common
+var _ cciptypes.ExtraDataCodecBundle = (*ExtraDataCodecRegistry)(nil)
+
+// ExtraDataCodecRegistry is a singleton registry that manages SourceChainExtraDataCodec instances
+// for different chain families. It implements the ExtraDataCodecBundle interface from chainlink-common
+// by delegating to the existing ExtraDataCodec implementation.
+//
+// Terminology:
+//   - "ExtraDataCodecRegistry": refers to the entire singleton registry instance. It both maintains the map of
+//     chain family to codec and provides thread-safe access to it.
+//   - "ExtraDataCodecBundle": is the interface defined in chainlink-common that the registry implements and that
+//     can be called over gRPC.
+type ExtraDataCodecRegistry struct {
+	extraDataCodec cciptypes.ExtraDataCodecMap
+	mu             sync.RWMutex
+}
+
+var (
+	registryInstance *ExtraDataCodecRegistry
+	registryOnce     sync.Once
+)
+
+// GetExtraDataCodecRegistry returns the singleton instance of ExtraDataCodecRegistry. This is only called
+// in core node.
+func GetExtraDataCodecRegistry() *ExtraDataCodecRegistry {
+	registryOnce.Do(func() {
+		registryInstance = &ExtraDataCodecRegistry{
+			extraDataCodec: make(cciptypes.ExtraDataCodecMap),
+		}
+	})
+	return registryInstance
+}
+
+// RegisterFamily registers a chain family with a no-op SourceChainExtraDataCodec if not already registered.
+// This is used when we know which chain families we want to support but don't have a specific codec
+// implementation initialized for it yet. This should only be called from core node, not over gRPC.
+func (r *ExtraDataCodecRegistry) RegisterFamily(chainFamily string) {
```

Comment:
wonder if it would be more accurate to use RegisterNoop(), and call it when codec is missing from config